### PR TITLE
Fixed remarks

### PR DESCRIPTION
**Pull request by cherhyshev**
Fix remarks of first implementation and add test for null and empty string.

### CODE REVIEW

File: src/main/java/ru/spbau/mit/DictionaryImpl.java [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    @Override
+    public int size() {
+        return notNullKeysSummary;
+    }
+
+    public int getBucketsSummary() {
+        return bucketsSummary;
+    }
+
+    private void setBucketsSummary(int newBucketSummary) {
```
Comment (dsavvinov) [inline src/main/java/ru/spbau/mit/DictionaryImpl.java]: Minor: нет большого смысла делать приватный сеттер -- сеттеры обычно нужны для того, чтобы скрыть (или иметь возможно скрыть в будущем) какие-то детали имплементации. От себя скрывать нечего, поэтому можно было просто прямо в поле и присваивать.

File: src/main/java/ru/spbau/mit/DictionaryImpl.java [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    @Override
+    public String get(@Nullable String key) {
+        int bucketIndex = getBucketIndex(key, getBucketsSummary());
+        DictionaryNode foundedNode = DictionaryNode.getNodeByKey(buckets[bucketIndex], key);
+        return (foundedNode == null) ? null : foundedNode.getValue();
+    }
+
+    @Override
+    public String put(@Nullable String key, @Nullable String value) {
+        if (key != null) {
```
Comment (dsavvinov) [inline src/main/java/ru/spbau/mit/DictionaryImpl.java]: В таких случаях лучше инвертировать условие: тогда одна ветка будет очень простой и понятной (если передали `null`, то вернули `null`), другая сложнее не станет, зато уменьшится горизонтальный отступ.

File: src/main/java/ru/spbau/mit/DictionaryImpl.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    public String remove(@Nullable String key) {
+        int bucketIndex = getBucketIndex(key, getBucketsSummary());
+        DictionaryNode containedNode = DictionaryNode.getNodeByKey(buckets[bucketIndex], key);
+        if (containedNode != null) {
+            buckets[bucketIndex] = DictionaryNode.removeByKey(buckets[bucketIndex], key);
+            decNotNullKeysSummary();
+
+            if (calcLoadFactor() < MIN_LOAD_FACTOR) {
+                int nextNumberOfBuckets = getBucketsSummary() / GROW_MEASURE;
+                if (DEFAULT_CAPACITY < nextNumberOfBuckets){
```
Comment (dsavvinov) [inline src/main/java/ru/spbau/mit/DictionaryImpl.java]: Это означает, что если у нас больше бакетов, чем `DEFAULT_CAPACITY`, но меньше, чем `2 * DEFAULT_CAPACITY`, то сокращения не произойдет. Не очень критично, но я бы все-таки ожидал, что будет сокращение до `min(nextNumberOfBuckets, DEFAULT_CAPACITY)`

File: src/main/java/ru/spbau/mit/DictionaryImpl.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+        setBucketsSummary(nextNumberOfBuckets);
+        buckets = newBuckets;
+    }
+
+
+    private int getBucketIndex(@Nullable String key, int bucketsSummary) {
+        return (key == null) ? 0 : Math.abs(key.hashCode()) % bucketsSummary;
+    }
+
+    private void incNotNullKeysSummary() {
```
Comment (dsavvinov) [inline src/main/java/ru/spbau/mit/DictionaryImpl.java]: То же самое замечание, что и про приватный сеттер -- если обращаться к полю напрямую, то никакой из принципов ООП не пострадает, а код станет попроще.

File: src/main/java/ru/spbau/mit/DictionaryImpl.java [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+    @Override
+    public String get(@Nullable String key) {
+        int bucketIndex = getBucketIndex(key, getBucketsSummary());
+        DictionaryNode foundedNode = DictionaryNode.getNodeByKey(buckets[bucketIndex], key);
+        return (foundedNode == null) ? null : foundedNode.getValue();
+    }
+
+    @Override
+    public String put(@Nullable String key, @Nullable String value) {
```
Comment (dsavvinov) [inline src/main/java/ru/spbau/mit/DictionaryImpl.java]: Что-то я не очень понимаю, какая политика в отношении null-ов.

По аннотациям кажется, что мы принимаем null и обещаем работать с ними.
По реализации метода `put` кажется, что мы `null` не принимаем, но без явного exception.
По реализации метода `remove` вообще непонятно -- вроде как будем пытаться искать `null` в первом бакете, но его там нет, т.к. `put` его не разрешает складывать.

File: src/main/java/ru/spbau/mit/DictionaryImpl.java [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+import org.jetbrains.annotations.Nullable;
+
+public final class DictionaryImpl implements Dictionary {
+    private static final int DEFAULT_CAPACITY = 5;
+    private static final double MAX_LOAD_FACTOR = 0.75;
+    private static final double MIN_LOAD_FACTOR = 0.25;
+    private static final int GROW_MEASURE = 2;
+
+    private DictionaryNode[] buckets = new DictionaryNode[DEFAULT_CAPACITY];
+    private int notNullKeysSummary;
```
[the remark this answers is not in the document]
Comment (cherhyshev) [reply src/main/java/ru/spbau/mit/DictionaryImpl.java]: fixed
